in_toto/verifylib.py

```diff
@@ -317,31 +330,54 @@ def verify_link_signature_thresholds(layout, chain_link_dict):
   # bad links, as long as we have enough good links. Only the good links will
   # be considered for further final product verification.
   for step in layout.steps:
-    # Contains all links corresponding to a step with successfully verified
-    # signatures, authorized to perform the step
+    # Will contain all links corresponding to a step with successfully
+    # verified signatures, authorized to perform the step
     verified_key_link_dict = {}
 
     # Iterate over links corresponding to a step
-    for keyid, link in six.iteritems(chain_link_dict.get(step.name, {})):
+    for link_keyid, link in six.iteritems(chain_link_dict.get(step.name, {})):
+      # Check if the link's keyid is authorized to provide a link for the step.
+      # Iterate over authorized keyids to find a key or subkey corresponding
+      # to the given link. Subkeys of authorized main keys are authorized
+      # implicitly.
+      for authorized_keyid in step.pubkeys:
+
+        authorized_key = layout.keys.get(authorized_keyid)
+        main_key_for_subkey = main_keys_for_subkeys.get(authorized_keyid)
+
+        # The signing key is authorized
+        if authorized_key and link_keyid == authorized_keyid:
+          verification_key = authorized_key
+          break
+
+        # The signing key is an authorized subkey
+        elif main_key_for_subkey and link_keyid == authorized_keyid:
+          verification_key = main_key_for_subkey
+          break
+
+        # The signing key is a subkey belonging to an authorized key
+        elif (authorized_key and
+            link_keyid in authorized_key.get("subkeys", {}).keys()):
+          verification_key = authorized_key
+          break
 
-      # Skip unauthorized links
-      if keyid not in step.pubkeys:
+      else:
         log.info("Skipping link. Keyid '{0}' is not authorized to sign links"
-            " for step '{1}'".format(keyid, step.name))
+            " for step '{1}'".format(link_keyid, step.name))
```

Comment:
Would it help to append the list of authorized keyids?

Reply:
I think it's okay that way.